src/common/buffer.cc

```diff
@@ -1757,7 +1757,59 @@ void buffer::list::decode_base64(buffer::list& e)
   push_back(std::move(bp));
 }
 
-  
+int buffer::list::pread_file(const char *fn, uint64_t off, uint64_t len, std::string *error)
+{
+  int fd = TEMP_FAILURE_RETRY(::open(fn, O_RDONLY|O_CLOEXEC));
+  if (fd < 0) {
+    int err = errno;
+    std::ostringstream oss;
+    oss << "can't open " << fn << ": " << cpp_strerror(err);
+    *error = oss.str();
+    return -err;
+  }
+
+  struct stat st;
+  memset(&st, 0, sizeof(st));
+  if (::fstat(fd, &st) < 0) {
+    int err = errno;
+    std::ostringstream oss;
+    oss << "bufferlist::read_file(" << fn << "): stat error: "
+        << cpp_strerror(err);
+    *error = oss.str();
+    VOID_TEMP_FAILURE_RETRY(::close(fd));
+    return -err;
+  }
+
+  if (off > st.st_size) {
+    std::ostringstream oss;
+    oss << "bufferlist::read_file(" << fn << "): read error: size < offset";
+    VOID_TEMP_FAILURE_RETRY(::close(fd));
+    return -1;
+  }
+
+  if (len > st.st_size - off) {
+    len = st.st_size - off;
+  }
+  lseek(fd, off, SEEK_SET);
+  ssize_t ret = safe_read(fd, (void*)this->c_str(), len);
```

Comment:
```suggestion
  ssize_t ret = read_fd(fd, len);
```
I don't think you should operate on `this->c_str()` directly...

Reply:
the use case here is to follow `pread(2)`. The caller need to allocate the `bufferlist` first, `bufferlist::pread_file` will try to read content from file and fill the `bufferlist`.

call `bufflist::read_fd(fd, len)` maybe not right here, as it will create a local buffer [first](https://github.com/ceph/ceph/blob/master/src/common/buffer.cc#L1760), read the content up to local buffer and push it back to the `bufferlist` user prepared. e.g. for a 4k read, the called prepared a 4k `bufferlist`, the read content will be pushed back to `bufferlist` - it will be a 8k `bufferlist` after `pread_file`

Reply:
> the called prepared a 4k `bufferlist`

I don't think the `bufferlist` is allocated before reading the files, so using `read_fd` here should get us a `bufferlist` with the desired contents and length. I used `gdb` to trace the length of `this->c_str()` before and after `read_fd(...)`, and the results did confirm this.

The reason I suggested this change is because `safe_read` causes the entire thing to crash on my environment. I am currently running a pretty heavily modified version of the immutable object cache daemon which uses the `ThreadPool` class for multithreading, and `safe_read(fd, (void*)this->c_str(), len)` always fails after several iterations because of "bad address (14)". I suspect the root cause of this issue is that multiple threads are trying to access the same `bufferlist`'s `c_str()`, but I could be wrong here.

Thanks!

Reply:
@usefulalgorithm  I just checked the code agin and see the issue now, you're right - I'm still in the impression of the previous `pread` based implementation. 
```
 76 int ObjectCacheFile::read_object_from_file(ceph::bufferlist* read_buf, uint64_t object_off, uint64_t object_len) {
 77
 78   ldout(cct, 20) << "offset:" << object_off
 79                  << ", length:" << object_len <<  dendl;
 80
 81   bufferptr buf(object_len);
 82
 83   // TODO(): aio
 84   int ret = pread(m_fd, buf.c_str(), object_len, object_off);
 85   if(ret < 0) {
 86     lderr(cct)<<"read file fail:" << std::strerror(errno) << dendl;
 87     return ret;
 88   }
 89   read_buf->append(std::move(buf));
 90
 91   return ret;
 92 }
```
 
Now with the current `bufferlist::pread_file`, we dont own pre-allocated buf, it's wrong to use `bufferlist::this->c_str()` which will return -1 and thus flow the read request to RADOS layer, instead we should generate this buf as you suggested.